src/pages/index.js

```diff
@@ -14,81 +16,190 @@ const addCardAddButton = document.querySelector(selectors.cardAddButton);
 updateProfilePopup.setEventListeners();
 
+// New card popup
+const newCardPopup = new PopupWithForm(selectors.cardPopup, handleAddCardFormSubmit);
+newCardPopup.setEventListeners();
+
+// Delete card popup
+const deletePopup = new PopupConfirm(selectors.deletePopup, handleDeleteCard);
+deletePopup.setEventListeners();
 
-function renderCard(item) {
-  const card = new Card(item, selectors.cardTemplate, handleImageClick);
-  return card.generateCard();
-}
+// Edit avatar popup
+const avatarPopup = new PopupWithForm(selectors.avatarPopup, handleEditAvatar);
+avatarPopup.setEventListeners();
 
-// ! ||--------------------------------------------------------------------------------||
-// ! ||                                 Event Listeners                                ||
-// ! ||--------------------------------------------------------------------------------||
+// Event Listeners
 addCardAddButton.addEventListener("click", () => {
   newCardPopup.open();
   cardFormValidator.toggleButtonState();
 });
 
 updateProfileButton.addEventListener("click", () => {
   const userData = userInfo.getUserInfo();
-  profileHeadingInput.value = userData.userName;
-  profileDescriptionInput.value = userData.userDescription;
+  profileHeadingInput.value = userData.name;
+  profileDescriptionInput.value = userData.about;
   editProfileFormValidator.resetValidation();
   updateProfilePopup.open();
 });
 
-// ! ||--------------------------------------------------------------------------------||
-// ! ||                                 Event Handlers                                 ||
-// ! ||--------------------------------------------------------------------------------||
-function handleAddCardFormSubmit(data) {
-  const card = renderCard(data);
-  cardSection.addItem(card);
+updateAvatarButton.addEventListener("click", () => {
+  avatarPopup.open();
+});
+
+// Functions
+function handleValidation(form) {
+  form.enableValidation();
+}
+
+function handleAddCardFormSubmit(cardData) {
+  newCardPopup.showButtonProgress(true);
+  api.addNewCard(cardData.name, cardData.link)
+    .then((res) => {
+      const card = renderCard(res);
+      cardSection.addItem(card);
+      newCardPopup.reset();
+      newCardPopup.close();
+    })
+    .catch((err) => {
+      console.error(`Error: ${err}`);
+    })
+    .finally(() => {
+      newCardPopup.showButtonProgress(false);
+    });
 }
 
 function handleImageClick(name, link) {
   newImagePopup.open(name, link);
 }
 
 function handleProfileSubmit(userData) {
-  userInfo.setUserInfo(userData);
-}
\ No newline at end of file
+  updateProfilePopup.showButtonProgress(true);
+  api.updateUserInfo(userData.name, userData.about)
+    .then((user) => {
+      userInfo.setUserInfo(user);
+      updateProfilePopup.reset();
+      updateProfilePopup.close();
+    })
+    .catch((err) => {
+      console.error(`Error: ${err}`);
+    })
+    .finally(() => {
+      updateProfilePopup.showButtonProgress(false);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```